### PR TITLE
NativeAOT-LLVM: Merge 31 jan 23

### PR DESCRIPTION
**Pull request by yowl**
Change `DllImport` to `LibraryImport` in several places

Inline `GetPointerType` for `ArrayType` `ArrayMethodKind.AddressWithHiddenArg`

Add
```
{ FUNC(CORINFO_HELP_READYTORUN_GCSTATIC_BASE) CORINFO_TYPE_PTR, { }, HFIF_SS_ARG },
{ FUNC(CORINFO_HELP_READYTORUN_NONGCSTATIC_BASE) CORINFO_TYPE_PTR, { }, HFIF_SS_ARG },
{ FUNC(CORINFO_HELP_READYTORUN_THREADSTATIC_BASE) CORINFO_TYPE_PTR, { }, HFIF_SS_ARG },
{ FUNC(CORINFO_HELP_READYTORUN_NONGCTHREADSTATIC_BASE) CORINFO_TYPE_PTR, { }, HFIF_SS_ARG },
```
Replacing
```
{ FUNC(CORINFO_HELP_READYTORUN_STATIC_BASE) CORINFO_TYPE_PTR, { }, HFIF_SS_ARG },
```

Remove use of `IsRuntimeExport`

Replace `factory.PreinitializationManager.HasLazyStaticConstructor(target)` with `TriggersLazyStaticConstructor` which does a `ConvertToCanonForm` on the `target`

Turn off hydration for LLVM/Wasm

Add some random value for `m_pPreviousTransitionFrame` to `AsmOffsetsCpu.h`

#ifdef out NUMASupportInitialize

Add fix for https://github.com/dotnet/runtimelab/issues/2252 : looks like an upstream issue, if you agree I will create the PR there.

### CODE REVIEW

**Comment by yowl** [comment]
Getting close with this, but have this question.  For the dump attached, I am hitting the assert at 
https://github.com/dotnet/runtimelab/blob/f9740bf8faae6ddcf576a802eed10a111b67f9e5/src/coreclr/jit/jithashtable.h#L238
with
```
>	clrjit_browser_wasm32_x64.dll!Llvm::buildStoreLocalVar(GenTreeLclVar * lclVar) Line 1283	C++
 	clrjit_browser_wasm32_x64.dll!Llvm::visitNode(GenTree * node) Line 1114	C++
 	clrjit_browser_wasm32_x64.dll!Llvm::generateBlock(BasicBlock * block) Line 354	C++
 	clrjit_browser_wasm32_x64.dll!`Llvm::generateBlocks'::`5'::LlvmCompileDomTreeVisitor::PreOrderVisit(BasicBlock * block) Line 322	C++
 	clrjit_browser_wasm32_x64.dll!DomTreeVisitor<`Llvm::generateBlocks'::`5'::LlvmCompileDomTreeVisitor>::WalkTree() Line 11326	C++
 	clrjit_browser_wasm32_x64.dll!Llvm::generateBlocks() Line 329	C++
 	clrjit_browser_wasm32_x64.dll!Llvm::Compile() Line 29	C++
 	clrjit_browser_wasm32_x64.dll!Compiler::compCompile::__l2::<lambda_6>::operator()() Line 5071	C++
 	clrjit_browser_wasm32_x64.dll!ActionPhase<`Compiler::compCompile'::`2'::<lambda_6>>::DoPhase() Line 65	C++
 	clrjit_browser_wasm32_x64.dll!Phase::Run() Line 61	C++
```

because the ssa local {3, 1} is already in the hash table by virtue of being initialised in the prolog:
```
Setting V03's initial value to uninit
  %2 = freeze i32 undef
```
I should note that I have removed the `fgDebugCheckSsa` because it was asserting on the `PostPhase` of the second SSA pass at `assert(fgDomsComputed);`  maybe I should be looking there?

cc @dotnet/nativeaot-llvm 

Edit:

Did a bit more investigation comparing with what happens before this merge and previously
```c++
bool hasImplicitDef = varDsc->GetPerSsaData(SsaConfig::FIRST_SSA_NUM)->GetAssignment() == nullptr;
```
finds an assignment so `hasImplicitDef ` is false nothing is done in the prolog for the local.  After the merge no assignment is found.


[localshashassert.txt.zip](https://github.com/dotnet/runtimelab/files/11205770/localshashassert.txt.zip)

File: eng/native/configureplatform.cmake [hunk 1]
```diff
@@ -459,7 +459,7 @@ else()
 endif(CLR_CMAKE_TARGET_UNIX)
 
 # check if host & target os/arch combination are valid
-if (NOT (CLR_CMAKE_TARGET_OS STREQUAL CLR_CMAKE_HOST_OS) AND NOT CLR_CMAKE_TARGET_WASI)
+if (NOT (CLR_CMAKE_TARGET_OS STREQUAL CLR_CMAKE_HOST_OS) AND NOT CLR_CMAKE_TARGET_WASI AND NOT CLR_CMAKE_TARGET_OS STREQUAL browser)
```
Comment (SingleAccretion) [inline eng/native/configureplatform.cmake]:
```suggestion
if (NOT (CLR_CMAKE_TARGET_OS STREQUAL CLR_CMAKE_HOST_OS) AND NOT CLR_CMAKE_TARGET_WASI AND NOT CLR_CMAKE_TARGET_BROWSER)
```
Or does it not work?

File: eng/native/gen-buildsys.cmd [hunk 2]
```diff
@@ -47,31 +36,22 @@ if /i "%__Ninja%" == "1" (
     if /i "%__Arch%" == "x86" (set __ExtraCmakeParams=%__ExtraCmakeParams% -A Win32)
 )
 
+echo checking arch
```
Comment (SingleAccretion) [inline eng/native/gen-buildsys.cmd]:
```suggestion
```

File: eng/native/gen-buildsys.cmd [hunk 3]
```diff
@@ -47,31 +36,22 @@ if /i "%__Ninja%" == "1" (
...
 if /i "%__Arch%" == "wasm" (
-<<<<<<< HEAD
-    set __ExtraCmakeParams=%__ExtraCmakeParams% "-DCMAKE_TOOLCHAIN_FILE=%EMSDK%/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake"
-    set __UseEmcmake=1
-=======
-
+    echo wasm
```
Comment (SingleAccretion) [inline eng/native/gen-buildsys.cmd]:
```suggestion
```

File: eng/native/gen-buildsys.cmd [hunk 4]
```diff
@@ -47,31 +36,22 @@ if /i "%__Ninja%" == "1" (
...
+            exit /B 1
         )
-        :: replace backslash with forward slash and append last slash
-        set "EMSDK_PATH=!EMSDK_PATH:\=/!"
-        if not "!EMSDK_PATH:~-1!" == "/" set "EMSDK_PATH=!EMSDK_PATH!/"
 
-        set __ExtraCmakeParams=%__ExtraCmakeParams% "-DCMAKE_TOOLCHAIN_FILE=!EMSDK_PATH!/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake"
+        set __ExtraCmakeParams=%__ExtraCmakeParams% "-DCMAKE_TOOLCHAIN_FILE=%EMSDK%/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake"
         set __UseEmcmake=1
+        echo using emcmake !__ExtraCmakeParams!
```
Comment (SingleAccretion) [inline eng/native/gen-buildsys.cmd]:
```suggestion
```

File: eng/native/gen-buildsys.cmd [hunk 5]
```diff
@@ -124,15 +103,19 @@ if not "%__ConfigureOnly%" == "1" (
     )
 )
 
+
+echo checkimg __UseEmcmake !__UseEmcmake!
```
Comment (SingleAccretion) [inline eng/native/gen-buildsys.cmd]:
```suggestion
```

File: eng/native/gen-buildsys.cmd [hunk 6]
```diff
@@ -124,15 +103,19 @@ if not "%__ConfigureOnly%" == "1" (
...
 if /i "%__UseEmcmake%" == "1" (
     REM workaround for https://github.com/emscripten-core/emscripten/issues/15440 - emscripten cache lock problems
     REM build the ports for ICU and ZLIB upfront
     embuilder build icu zlib
 
     REM Add call in front of emcmake as for some not understood reason, perhaps to do with scopes, by calling emcmake (or any batch script),
     REM delayed expansion is getting turned off. TODO: remove this and see if CI is ok and hence its just my machine.
+echo calling emcmake
```
Comment (SingleAccretion) [inline eng/native/gen-buildsys.cmd]:
```suggestion
```

File: eng/native/gen-buildsys.cmd [hunk 7]
```diff
@@ -124,15 +103,19 @@ if not "%__ConfigureOnly%" == "1" (
...
     call emcmake "%CMakePath%" %__ExtraCmakeParams% --no-warn-unused-cli -G "%__CmakeGenerator%" -B %__IntermediatesDir% -S %__SourceDir% 
 setlocal EnableDelayedExpansion EnableExtensions
+echo called emcmake
```
Comment (SingleAccretion) [inline eng/native/gen-buildsys.cmd]:
```suggestion
```

File: src/coreclr/CMakeLists.txt [hunk 8]
```diff
@@ -145,11 +145,7 @@ endif(NOT CLR_CMAKE_TARGET_ARCH_WASM)
 
 if(NOT CLR_CROSS_COMPONENTS_BUILD)
   # NativeAOT only buildable for a subset of CoreCLR-supported configurations
-<<<<<<< HEAD
-  if((CLR_CMAKE_HOST_LINUX OR CLR_CMAKE_HOST_OSX OR CLR_CMAKE_HOST_WIN32 OR EMSCRIPTEN) AND (CLR_CMAKE_HOST_ARCH_ARM64 OR CLR_CMAKE_HOST_ARCH_AMD64 OR CLR_CMAKE_TARGET_ARCH_WASM))
-=======
-  if(CLR_CMAKE_HOST_ARCH_ARM64 OR CLR_CMAKE_HOST_ARCH_AMD64)
->>>>>>> d7d154d7e25b5a4472b75c963b0a73dc23f5fb9b
+  if(CLR_CMAKE_HOST_ARCH_ARM64 OR CLR_CMAKE_HOST_ARCH_AMD64 OR (EMSCRIPTEN AND CLR_CMAKE_TARGET_ARCH_WASM))
```
Comment (SingleAccretion) [inline src/coreclr/CMakeLists.txt]:
```suggestion
  if(CLR_CMAKE_HOST_ARCH_ARM64 OR CLR_CMAKE_HOST_ARCH_AMD64 OR CLR_CMAKE_TARGET_ARCH_WASM)
```

File: src/coreclr/jit/ee_il_dll.cpp [hunk 9]
```diff
@@ -135,7 +130,9 @@ void jitprintf(const char* fmt, ...)
 }
 #endif
 
->>>>>>> d7d154d7e25b5a4472b75c963b0a73dc23f5fb9b
+#if TARGET_WASM
+extern "C" DLLEXPORT
+#endif
 void jitShutdown(bool processIsTerminating)
```
Comment (SingleAccretion) [inline src/coreclr/jit/ee_il_dll.cpp]: Is this export still needed?

File: src/coreclr/jit/fgdiagnostic.cpp [hunk 10]
```diff
@@ -3542,7 +3542,9 @@ void Compiler::fgDebugCheckLinks(bool morphTrees)
     }
 
     fgDebugCheckNodesUniqueness();
+#ifndef TARGET_WASM // hits assert(fgDomsComputed);
```
Comment (SingleAccretion) [inline src/coreclr/jit/fgdiagnostic.cpp]:
```suggestion
#ifndef TARGET_WASM // Does not handle LIR.
```

File: src/coreclr/jit/gentree.cpp [hunk 11]
```diff
@@ -17316,8 +17316,14 @@ bool GenTree::IsPhiNode()
 
 bool GenTree::IsPhiDefn()
 {
+#ifdef TARGET_WASM // LIR SSA pass can have GT_STORE_LCL_VAR as a phi arg
+    bool res = OperIs(GT_ASG) && AsOp()->gtOp2->OperIs(GT_PHI) ||
+               ((OperGet() == GT_STORE_LCL_VAR) && (AsOp()->gtOp1 != nullptr) && (AsOp()->gtOp1->OperGet() == GT_PHI));
```
Comment (SingleAccretion) [inline src/coreclr/jit/gentree.cpp]:
```suggestion
               ((OperGet() == GT_STORE_LCL_VAR) && (AsOp()->gtOp1->OperGet() == GT_PHI));
```

File: src/coreclr/jit/llvmdebuginfo.cpp [hunk 12]
```diff
@@ -234,9 +234,11 @@ void Llvm::declareDebugVariables()
     DILocation* debugLocation = getArtificialDebugLocation();
     Instruction* insertInst = _builder.GetInsertBlock()->getTerminator();
     Value* spilledShadowStackAddr = nullptr;
-    for (auto lcl = m_debugVariablesMap.Begin(); !lcl.Equal(m_debugVariablesMap.End()); ++lcl)
+    for (auto lcl :
+         JitHashTable<unsigned, JitSmallPrimitiveKeyFuncs<unsigned>, llvm::DILocalVariable*>::KeyValueIteration(
+             &m_debugVariablesMap))
```
Comment (SingleAccretion) [inline src/coreclr/jit/llvmdebuginfo.cpp]:
```suggestion
    for (auto lcl : decltype(m_debugVariablesMap)::KeyValueIteration(&m_debugVariablesMap))
```
`decltype` can be useful in situations like this.

File: src/coreclr/jit/ssabuilder.cpp [hunk 13]
```diff
@@ -852,68 +853,26 @@ void SsaBuilder::RenameDef(GenTree* defNode, BasicBlock* block)
...
     bool                 isLocal   = defNode->DefinesLocal(m_pCompiler, &lclNode, &isFullDef, &offset, &storeSize);
->>>>>>> d7d154d7e25b5a4472b75c963b0a73dc23f5fb9b
+#endif
 
     if (isLocal)
     {
+#ifndef TARGET_WASM
         // This should have been marked as definition.
         assert(((lclNode->gtFlags & GTF_VAR_DEF) != 0) && (((lclNode->gtFlags & GTF_VAR_USEASG) != 0) == !isFullDef));
+#endif // TARGET_WASM
```
Comment (SingleAccretion) [inline src/coreclr/jit/ssabuilder.cpp]: This assert should (must) hold for LIR as well. If there are cases where it doesn't they should be fixed.

File: src/coreclr/jit/ssabuilder.cpp [hunk 14]
```diff
@@ -852,68 +853,26 @@ void SsaBuilder::RenameDef(GenTree* defNode, BasicBlock* block)
...
-        isLocal = defNode->DefinesLocal(m_pCompiler, &lclNode, &isFullDef);
+        isLocal = defNode->DefinesLocal(m_pCompiler, &lclNode, &isFullDef, &offset, &storeSize);
     }
 #else
-    bool isLocal = defNode->DefinesLocal(m_pCompiler, &lclNode, &isFullDef);
-#endif
-=======
-    ssize_t              offset    = 0;
-    unsigned             storeSize = 0;
     bool                 isLocal   = defNode->DefinesLocal(m_pCompiler, &lclNode, &isFullDef, &offset, &storeSize);
```
Comment (SingleAccretion) [inline src/coreclr/jit/ssabuilder.cpp]:
```suggestion
    bool isLocal = defNode->DefinesLocal(m_pCompiler, &lclNode, &isFullDef, &offset, &storeSize);
```

File: src/coreclr/jit/ssabuilder.cpp [hunk 15]
```diff
@@ -947,8 +906,20 @@ void SsaBuilder::RenameDef(GenTree* defNode, BasicBlock* block)
                         lclNode->SetSsaNum(m_pCompiler, index, ssaNum);
                     }
                 }
+#ifdef TARGET_WASM // TODO-LLVM: clearing the SSA number for locals not in SSA, can we remove the IsLIR check?
+                else if (block->IsLIR())
+                {
+                    lclNode->SetSsaNum(m_pCompiler, index, SsaConfig::RESERVED_SSA_NUM);
+                }
+#endif
```
Comment (SingleAccretion) [inline src/coreclr/jit/ssabuilder.cpp]: We should fix `fgResetForSsa` to handle LIR instead.

File: src/coreclr/jit/ssabuilder.cpp [hunk 16]
```diff
@@ -947,8 +906,20 @@ void SsaBuilder::RenameDef(GenTree* defNode, BasicBlock* block)
...
             }
         }
+#ifdef TARGET_WASM // TODO-LLVM: clearing the SSA number for locals not in SSA, can we remove the IsLIR check?
+        else if (block->IsLIR())
+        {
+            lclNode->SetSsaNum(SsaConfig::RESERVED_SSA_NUM);
+        }
+#endif
```
Comment (SingleAccretion) [inline src/coreclr/jit/ssabuilder.cpp]: Same point about `fgResetForSsa`.

File: src/coreclr/jit/ssabuilder.cpp [hunk 17]
```diff
@@ -1038,8 +1009,11 @@ unsigned SsaBuilder::RenamePushDef(GenTree* defNode, BasicBlock* block, unsigned
     assert(m_pCompiler->lvaInSsa(lclNum) && !m_pCompiler->lvaGetDesc(lclNum)->lvPromoted);
 
     LclVarDsc* const varDsc = m_pCompiler->lvaGetDesc(lclNum);
-    unsigned const   ssaNum =
-        varDsc->lvPerSsaData.AllocSsaNum(m_allocator, block, defNode->OperIs(GT_ASG) ? defNode->AsOp() : nullptr);
+    unsigned const   ssaNum = varDsc->lvPerSsaData.AllocSsaNum(m_allocator, block, defNode->OperIs(GT_ASG)
+#ifdef TARGET_WASM // for the second pass through SSA with LIR
+            || defNode->OperIs(GT_STORE_LCL_VAR)
+#endif // TARGET_WASM
+            ? defNode->AsOp() : nullptr);
```
Comment (SingleAccretion) [inline src/coreclr/jit/ssabuilder.cpp]:
```suggestion
    unsigned const   ssaNum =
        varDsc->lvPerSsaData.AllocSsaNum(m_allocator, block, !defNode->IsCall() ? defNode->AsOp() : nullptr);
```
I should've written it this way originally...

File: src/coreclr/jit/treelifeupdater.cpp [hunk 18]
```diff
@@ -51,13 +51,10 @@ bool TreeLifeUpdater<ForCodeGen>::UpdateLifeFieldVar(GenTreeLclVar* lclNode, uns
...
     if (isBorn || isDying)
     {
-<<<<<<< HEAD
-#ifndef TARGET_WASM
-=======
         bool previouslyLive = VarSetOps::IsMember(compiler, compiler->compCurLife, fldVarDsc->lvVarIndex);
         UpdateLifeBit(compiler->compCurLife, fldVarDsc, isBorn, isDying);
 
->>>>>>> d7d154d7e25b5a4472b75c963b0a73dc23f5fb9b
+#ifndef TARGET_WASM // this method is never called for TARGET_WASM, just #if-ing out code to make it compile
```
Comment (SingleAccretion) [inline src/coreclr/jit/treelifeupdater.cpp]:
```suggestion
#ifndef TARGET_WASM // Not used and not compilable on WASM.
```

File: src/coreclr/jit/treelifeupdater.cpp [hunk 19]
```diff
@@ -329,6 +243,7 @@ void TreeLifeUpdater<ForCodeGen>::UpdateLifeVar(GenTree* tree, GenTreeLclVarComm
                     continue;
                 }
 
+#ifndef TARGET_WASM // the rest of this code will never get reached as ForCodeGen is always false for TARGET_WASM
```
Comment (SingleAccretion) [inline src/coreclr/jit/treelifeupdater.cpp]:
```suggestion
#ifndef TARGET_WASM // Not reachable and not compilable on WASM.
```

File: src/coreclr/nativeaot/Runtime/CMakeLists.txt [hunk 20]
```diff
@@ -228,20 +228,15 @@ if (NOT CLR_CMAKE_TARGET_ARCH_WASM)
...
-<<<<<<< HEAD
 
 if(CLR_CMAKE_TARGET_UNIX)
-  add_definitions(-DFEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP)
   
   # FEATURE_MANUALLY_MANAGED_CARD_BUNDLES only supported for 64 bit hosts
   if (NOT CLR_CMAKE_TARGET_ARCH_WASM) 
     add_definitions(-DFEATURE_MANUALLY_MANAGED_CARD_BUNDLES)
   endif()
 endif()
```
Comment (SingleAccretion) [inline src/coreclr/nativeaot/Runtime/CMakeLists.txt]: It looks like `DFEATURE_MANUALLY_MANAGED_CARD_BUNDLES` was enabled on Windows as well, so this whole section should be:
```
if (NOT CLR_CMAKE_TARGET_ARCH_WASM) # FEATURE_MANUALLY_MANAGED_CARD_BUNDLES only supported for 64 bit hosts
  add_definitions(-DFEATURE_MANUALLY_MANAGED_CARD_BUNDLES)
endif()
```
(After `add_definitions(-DFEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP)`)

File: src/tests/run.py [hunk 21]
```diff
@@ -1097,6 +1097,8 @@ def match_filename(test_path):
         if not os.path.isdir(os.path.dirname(test_path)):
             pass
 
+        print("test_path")
+        print(test_path)
```
Comment (SingleAccretion) [inline src/tests/run.py]:
```suggestion
```

File: eng/native/configureplatform.cmake (hunk 1 above)
Comment (yowl) [reply eng/native/configureplatform.cmake]: it does work, thanks

File: src/coreclr/jit/ssabuilder.cpp (hunk 13 above)
Comment (yowl) [reply src/coreclr/jit/ssabuilder.cpp]: ok, I hit this with a `0x0000027d84d9d4c8 135: [[GT_STORE_LCL_FLD, TYP_STRUCT]` where the flags are `8000 000D` i.e.
`GTF_VAR_DEF` (good) but not `GTF_VAR_USEASG` when `isFullDef  == false`.  Do you want the full dump?

File: src/coreclr/jit/ee_il_dll.cpp (hunk 9 above)
Comment (yowl) [reply src/coreclr/jit/ee_il_dll.cpp]: No, deleted.

File: src/coreclr/jit/ssabuilder.cpp (hunk 13 above)
Comment (yowl) [reply src/coreclr/jit/ssabuilder.cpp]: I have commited a change for this, basically ive just taken the logic that is in the assert and used it to set `GTF_VAR_USEASG`.

File: src/coreclr/jit/ssabuilder.cpp [hunk 22]
```diff
@@ -115,10 +115,24 @@ void Compiler::fgResetForSsa()
...
+                    {
+                        for (unsigned index = 0; index < varDsc->lvFieldCnt; index++)
+                        {
+                            unsigned   fieldLclNum = varDsc->lvFieldLclStart + index;
+                            if (!lvaInSsa(fieldLclNum))
+                            {
+                                lclNode->SetSsaNum(this, index, SsaConfig::RESERVED_SSA_NUM);
+                            }
+                        }
+                    }
```
Comment (SingleAccretion) [inline src/coreclr/jit/ssabuilder.cpp]: Hmm, it is very surprising this is needed. `SetSsaNum(SsaConfig::RESERVED_SSA_NUM);` should be enough to clear out the SSA number, composite or not.

**Comment by yowl** [comment]
Sorry, I left a lot of debug there.  Have cleared up the remaining feedback `fgResetForSsa` and set `GTF_VAR_USEASG` in a couple of additional places to renable the assert.  @SingleAccretion

File: src/coreclr/jit/ssabuilder.cpp [hunk 23]
```diff
@@ -840,7 +854,8 @@ void SsaBuilder::RenameDef(GenTree* defNode, BasicBlock* block)
 
     GenTreeLclVarCommon* lclNode;
     bool                 isFullDef = false;
-<<<<<<< HEAD
+    ssize_t              offset    = 0;
+    unsigned             storeSize = 0;
 #if defined(TARGET_WASM)
```
Comment (SingleAccretion) [inline src/coreclr/jit/ssabuilder.cpp:844]: We should also set these in the `block->IsLIR()` branch (for consistency - they won't be used now as we don't expect composite definitions in late SSA):
```
if (block->IsLIR())
{
    isLocal = defNode->OperIsLocalStore();
    lclNode = isLocal ? defNode->AsLclVarCommon() : nullptr;
    if (isLocal)
    {
        isFullDef = !lclNode->IsPartialLclFld(m_pCompiler);
        offset    = lclNode->GetLclOffs();
        storeSize = isFullDef ? m_pCompiler->lvaLclExactSize(lclNode->GetLclNum()) : lclNode->AsLclFld()->GetSize();
    }
}
```

File: src/coreclr/jit/llvmlower.cpp [hunk 24]
```diff
@@ -670,6 +670,12 @@ void Llvm::lowerLocal(GenTreeLclVarCommon* node)
     }
 }
 
+static bool isFullDefStore(GenTreeLclVarCommon* lclVar, LclVarDsc* varDsc)
```
Comment (SingleAccretion) [inline src/coreclr/jit/llvmlower.cpp]: I don't think there is a need for the new helper: `IsPartialLclFld` will suffice.

File: src/coreclr/jit/llvmlower.cpp [hunk 25]
```diff
@@ -736,8 +750,10 @@ void Llvm::lowerFieldOfDependentlyPromotedStruct(GenTree* node)
             if ((node->gtFlags & GTF_VAR_DEF) != 0)
             {
                 // Conservatively assume these become partial.
```
Comment (SingleAccretion) [inline src/coreclr/jit/llvmlower.cpp]:
```suggestion
```

File: src/tests/run.py (hunk 21 above)
Comment (SingleAccretion) [reply src/tests/run.py]: This one was missed?

File: src/coreclr/jit/ssabuilder.cpp (hunk 22 above)
Comment (yowl) [reply src/coreclr/jit/ssabuilder.cpp]: ah, you are right.  I don't need it now.  Will remove.

File: src/coreclr/jit/ssabuilder.cpp (hunk 23 above)
Comment (yowl) [reply src/coreclr/jit/ssabuilder.cpp:844]: ok, have added :
```c++
storeSize = m_pCompiler->lvaLclExactSize(lclNode->GetLclNum());

if (lclNode->OperIs(GT_STORE_LCL_FLD))
{
    offset = lclNode->AsLclFld()->GetLclOffs();
}
else
{
    offset = lclNode->GetLclOffs();
}
```

File: src/coreclr/jit/llvmlower.cpp (hunk 24 above)
Comment (yowl) [reply src/coreclr/jit/llvmlower.cpp]: I have a problem with `IsPartialLclFld` when `TypeIs(TYP_STRUCT)` as the size comparison is different.  Can struct stores be `GTF_VAR_USEASG` I suppose they can?

Comment (SingleAccretion) [reply src/coreclr/jit/llvmlower.cpp]: Yes, they can.

Did you also change the code in `ssabuilder.cpp` to use `IsPartialLclFld` (they have to stay in sync)?

File: src/coreclr/jit/ssabuilder.cpp (hunk 23 above)
Comment (SingleAccretion) [reply src/coreclr/jit/ssabuilder.cpp:844]: > `storeSize = m_pCompiler->lvaLclExactSize(lclNode->GetLclNum());`

That won't be correct for `STORE_LCL_FLD`. For `offset` - `GenTreeLclVarCommon::GetLclOffs` didn't work?

Comment (yowl) [reply src/coreclr/jit/ssabuilder.cpp:844]: Don't fully understand the second sentence, but have changed to 
```c++
if (lclNode->OperIs(GT_STORE_LCL_FLD))
{
    offset = lclNode->AsLclFld()->GetLclOffs();
    storeSize = lclNode->AsLclFld()->GetSize();
}
```
to match `DefinesLocal` for `GT_LCL_FLD`

File: src/coreclr/jit/llvmlower.cpp (hunk 24 above)
Comment (yowl) [reply src/coreclr/jit/llvmlower.cpp]: No I did not, thanks!  Lets see if it goes green.

File: src/coreclr/jit/ssabuilder.cpp (hunk 23 above)
Comment (SingleAccretion) [reply src/coreclr/jit/ssabuilder.cpp:844]: > Don't fully understand the second sentence

I meant that `offset = lclNode->GetLclOffs();` should work as-is, there is no need for the `if`.

Comment (yowl) [reply src/coreclr/jit/ssabuilder.cpp:844]: How about:
```c++
offset = lclNode->GetLclOffs();

if (lclNode->OperIs(GT_STORE_LCL_FLD))
{
    storeSize = lclNode->AsLclFld()->GetSize();
}
else
{
    storeSize = m_pCompiler->lvaLclExactSize(lclNode->GetLclNum());
}
```

Comment (SingleAccretion) [reply src/coreclr/jit/ssabuilder.cpp:844]: LGTM.

File: src/coreclr/jit/ssabuilder.cpp [hunk 26]
```diff
@@ -840,28 +839,34 @@ void SsaBuilder::RenameDef(GenTree* defNode, BasicBlock* block)
...
+        isFullDef = !lclNode->IsPartialLclFld(m_pCompiler);
+
+        if (lclNode->OperIs(GT_STORE_LCL_FLD))
+        {
+            offset = lclNode->AsLclFld()->GetLclOffs();
+            storeSize = lclNode->AsLclFld()->GetSize();
+        }
+        else
+        {
+            offset = lclNode->GetLclOffs();
```
Comment (SingleAccretion) [inline src/coreclr/jit/ssabuilder.cpp]:
```suggestion
            offset = 0;
```

**Review by SingleAccretion — APPROVED** [review]
LGTM, thank you!

**Comment by yowl** [comment]
Thanks for the review, one more to go and we will be up to date, at least for a while!

File: src/coreclr/jit/ssabuilder.cpp (hunk 26 above)
Comment (jkotas) [reply src/coreclr/jit/ssabuilder.cpp]: @yowl Would you like to accept this suggested change?

File: src/coreclr/jit/ssabuilder.cpp (hunk 23 above)
Comment (yowl) [reply src/coreclr/jit/ssabuilder.cpp:844]: oops, didn't actually commit this, have now.

File: src/coreclr/jit/ssabuilder.cpp (hunk 26 above)
Comment (yowl) [reply src/coreclr/jit/ssabuilder.cpp]: ah, we've got crossed somewhere, I've added another commit to match the suggestion

**Pull request merged**